test/tests/functional/pbs_snapshot_unittest.py

```diff
@@ -97,9 +98,17 @@ def setup_sc(self, sched_id, partition, port):
         a = {'partition': partition,
              'sched_host': self.server.hostname,
              'sched_port': port}
+        if sched_priv is not None:
+            a.update({'sched_priv': sched_priv})
+        if sched_log is not None:
+            a.update({'sched_log': sched_log})
         self.server.manager(MGR_CMD_CREATE, SCHED, a, id=sched_id)
         self.scheds[sched_id].create_scheduler()
-        self.scheds[sched_id].start()
+        if(len(a) > 3):
+            sched_dir = os.path.dirname(sched_priv)
+            self.scheds[sched_id].start(sched_dir)
```

Comment:
I took a closer look at Scheduler.start(sched_home), it seems like 'sched_home' is passed as PBS_HOME argument to pbs_sched, which seems wrong to me. @suresh-thelkar, @arungrover and @bhroam , can you guys please help me understand if this is a bug? if 'sched_home', which is parent directory of a sched's sched_priv/log, is set as PBS_HOME for that sched, then it'd imply that the sched will search for this parent directory for the existence of the job_sort_formula file (instead of PBS_HOME) right? Won't this be an issue? job_sort_formula should only exist inside a file in PBS_HOME right?

Reply:
@agrawalravi90,

Following are my thoughts on this.

We need to pass sched_home directory to  Scheduler.start() only when we create scheduler's priv and log directories outside the regular PBS_HOME.  This parameter serves as value for -d argument of pbs_sched.  If scheduler's priv and log directories are inside PBS_HOME then we no need to pass value for sched_home.

Each multisched having its own job_sort_formula is not supported at the moment. We are ready with our design/analysis to for this enhancement which is not getting prioritized.

Reply:
Thanks for your reply Suresh.

I was not thinking about each multisched having its own job_sort_formula, I was concerned about what would happen if we set job_sort_formula on the server with a sched that was run with -d /tmp (for example) ? Wouldn't that sched try to access the sched_formula file inside /tmp/sched_priv? Will it find a copy of the job_sort_formula's file (sched_formula?) there?

Reply:
@suresh-thelkar Was that fact that job_sort_formula didn't work with multi-sched clear in your EDD?  This needs to be communicated in the docs or release notes properly.

Reply:
@bhroam,
As we have not yet made job_sort_formula a scheduler attribute we initially thought of not adding it to the EDD. An enhancement to make job_sort_formula to work with Multisched has been logged already. We would like to add a new interface to Multisched EDD once we start working on this enhancement. 
Since scheduler's priv and log directories can be created in any path, planning to add the following caveat either to the documentation/release notes.  Thanks for reminding us. Hope the following information is enough. Please let me know your thoughts in case if we need to add more details.
**Caveat**: Each multisched uses/shares the same job_sort_formula set at server for the time being. Scheduler's sched_priv and sched_log directories should be at the same level as of default scheduler's priv and log directories as of now.

Hope @agrawalravi90 the above explanation answers your question as well. Please let us know in case if anything more needs to be added.

Reply:
@suresh-thelkar The scheduler doesn't read the job_sort_formula attribute from the server.  It is written into a file in sched_priv.  The scheduler reads it from the file.  This is for security reasons.  We don't want a non-root manager to set this attribute and have arbitrary python code run as root.  From what @agrawalravi90 just said, the file isn't being copied into the non-default scheduler's sched_priv directories.  This means the job_sort_formula isn't shared by all schedulers.  It will only work for the default scheduler.

Reply:
Thanks for letting us know the intent behind copying the sched_formula to sched_priv. Thereafter this file is used by scheduler to read the formula and proceed accordingly.  I am still thinking that job_sort_formula works for Multisched as well( with the same formula as that of default scheduler and with the above caveat I mentioned in the above comments).  In order to find more details I have executed the scenario attached to confirm whether the same formula is used for both default scheduler as well as Multisched. It looks to me that both are using the formula and jobs are sorted accordingly.  Please look into the attachment and correct me If I am missing anything.
[Existing_Job_sort_formula_Analysis.txt](https://github.com/PBSPro/pbspro/files/2066838/Existing_Job_sort_formula_Analysis.txt)

Also looked at the code to see how scheduler is reading the job_sort_formula. It calls the function read_formula() where the scheduler file is read from the macro FORMULA_ATTR_PATH_SCHED which is nothing but "sched_priv/sched_formula". So even in case of Multisched also since this code has not been changed as of now it still reads the formula from the same macro(only thing is that corresponding Multisched should have access to default scheduler's sched_priv directory). So with respect to this I am feeling that job_sort_formula works for Multisched as well except with the following caveat till the time we complete the job_sort_formula enhancement for Multisched.

**Caveat**: Each multisched tries to use the same job_sort_formula of default scheduler for the time being. As long as Multisched's sched_priv_<sc_name> directory is created at the same level as that of default scheduler's sched_priv directory.
Example: sched_priv_<sc_name> is strictly created in PBS_HOME.

Reply:
Please have this caveat added to the release notes for multi-sched.

Reply:
Thanks for clarifying Suresh. I'm sorry but I have one more question: Right now, pbs_sched -d has been overloaded to accept the parent directory of sched_priv and sched_log. So, what will happen if they are put in different locations? Or is that not supported? Or is pbs_sched -d not supported? I couldn't find any references to pbs_sched -d in the EDD.